### DISCUSSION TITLE
Better Initial Guess for Newton Solve

### DISCUSSION THREAD

**Discussion by maxnezdyur** [idaholab/moose#22319] [Q&A General] [answered]
I looked at the predictor system and I see that for implicit solves it can give a better initial guess for the first Newton Iteration. Say I know that for a specific dof, there is a better initial guess. Is there a way for me to supply that dof that guess but keep the rest the same?

**Comment by GiudGiud** [comment, marked as answer]
We don’t usually modify initial guesses of each time step solves, except the first one, because the initial guess is the solution from the previous step. ~~and the time derivative term will also depend on it. So if you modify the “initial guess” you modify the solution of the transient~~

~~You ll have to write a custom user object to do that.~~
See below for custom predictor

**Reply by maxnezdyur** [reply]
When using the predictor block for an implicit solve, I thought that it didn't change the final solution just the first guess? I am bascially trying to implement that but with a single dof. Would that still modify the final solution?

**Reply by GiudGiud** [reply]
It must be doing something else to get the time derivative right. 
I m familiar with using a predictor in a predictor corrector scheme, not on its own.

**Reply by hugary1995** [reply]
> So if you modify the “initial guess” you modify the solution of the transient

The predictor modifies the current solution, not the _solution_old. So you don't need to do anything special to get the time derivative right.

**Reply by hugary1995** [reply]
@maxnezdyur, you can derive from `SimplePredictor`, and if you know the dof you want to modify, modify it in `MyCustomPredictor::apply(NumericVector<Number> & sln)`. E.g.

```cpp
void
MyCustomPredictor::apply(NumericVector<Number> & sln)
{
  SimplePredictor::apply(sln);

  // whatever it takes for you to figure out that dof and the modified guess.
  ......

  sln.set(dof, value);
}
```

**Reply by hugary1995** [reply]
... or just derive from `Predictor` if `SimplePredictor` isn't what you want.

**Reply by maxnezdyur** [reply]
The final question today (thanks for all the help everyone). Is there a way to get the nodal neighbors of a node? I am currently using a nodal user object, and have access to the node and mesh, but I can't find a way to find the nodes connected to the current node.

**Reply by hugary1995** [reply]
You can request a node-to-elem-map. I assume by "nodes connected to the current node" you mean nodes of the elements that are connected to the current node.

**Comment by friedmud** [comment]
If you are doing a Steady solve... then the initial guess is just the initial condition... so you can use that system to set an initial guess.

@hugary1995 has the right idea for Transient solves.

Also, for BCs, note that the `DirichletBC` object is always doing this (modifying the initial guess to match the value) by default (you can turn it off by setting `preset = false`).
